wrappers/golang/curves/bls12377/poseidon/poseidon.go

```diff
@@ -0,0 +1,57 @@
+package poseidon
+
+// #cgo CFLAGS: -I./include/
+// #include "poseidon.h"
+import "C"
+
+import (
+	"unsafe"
+
+	"github.com/ingonyama-zk/icicle/v2/wrappers/golang/core"
+	cr "github.com/ingonyama-zk/icicle/v2/wrappers/golang/cuda_runtime"
+)
+
+func GetDefaultPoseidonConfig() core.PoseidonConfig {
+	return core.GetDefaultPoseidonConfig()
+}
+
+func PoseidonHash[T any](scalars, results core.HostOrDeviceSlice, numberOfStates, arity int, cfg *core.PoseidonConfig, constants *core.PoseidonConstants[T]) core.IcicleError {
+	scalarsPointer, resultsPointer, cfgPointer := core.PoseidonCheck(scalars, results, cfg, constants, arity, numberOfStates)
+
+	cScalars := (*C.scalar_t)(scalarsPointer)
+	cResults := (*C.scalar_t)(resultsPointer)
+	cNumberOfStates := (C.int)(numberOfStates)
+	cArity := (C.int)(arity)
+	cConstants := (*C.PoseidonConstants)(unsafe.Pointer(constants))
+	cCfg := (*C.PoseidonConfig)(cfgPointer)
+
+	__ret := C.bls12_377_poseidon_hash_cuda(cScalars, cResults, cNumberOfStates, cArity, cConstants, cCfg)
+
+	err := (cr.CudaError)(__ret)
+	return core.FromCudaError(err)
+}
+
+func CreateOptimizedPoseidonConstants[T any](arity, fullRoundsHalfs, partialRounds int, constants core.HostOrDeviceSlice, ctx cr.DeviceContext, poseidonConstants *core.PoseidonConstants[T]) core.IcicleError {
+
+	cArity := (C.int)(arity)
+	cFullRoundsHalfs := (C.int)(fullRoundsHalfs)
+	cPartialRounds := (C.int)(partialRounds)
+	cConstants := (*C.scalar_t)(constants.AsUnsafePointer())
+	cCtx := (*C.DeviceContext)(unsafe.Pointer(&ctx))
+	cPoseidonConstants := (*C.PoseidonConstants)(unsafe.Pointer(poseidonConstants))
+
+	__ret := C.bls12_377_create_optimized_poseidon_constants_cuda(cArity, cFullRoundsHalfs, cPartialRounds, cConstants, cCtx, cPoseidonConstants)
+	err := (cr.CudaError)(__ret)
+	return core.FromCudaError(err)
+}
```

Comment:
@ChickenLover Whats the use case for calling this instead of `InitOptimizedPoseidonConstantsCuda`?

Reply:
made a binding because there was an extern link. figured there's a use case for it

Reply:
@jeremyfelder `create` allows user to pass his own constants. While `Init` just loads pre-defined ones